Use new native text-buffer implementation

maxbrunsfeld opened the pull request:
### Summary

This pull request completely reimplements the `TextBuffer` class - Atom's core data structure representing editable text. The new implementation differs from the old one in several major ways:

* The class is now implemented primarily in C++, as part of our native library [superstring](https://github.com/atom/superstring).
* Instead of storing the text as a mutable array of lines, we store it as flat immutable string representing the file's original content, plus a *Patch* representing the user's unsaved changes. This is somewhat similar to the [piece table](https://en.wikipedia.org/wiki/Piece_table) data structure used in some other editors.
* At the native level, the buffer supports *snapshots* - read-only views of the buffer that can be processed in background threads with no copying or locking, even while the buffer is being mutated on the main thread.
* File I/O and transcoding is now handled by the native library, entirely on background threads and exposed via promises.

### Motivation

For many purposes, the existing array-of-lines implementation was actually very fast; we've never observed slowness associated with the querying or mutating of the text itself. This is a testament to V8's highly efficient String operations. However, there are many operations in Atom that *use* the `TextBuffer` and would benefit from being dropped to C++ and/or being moved to a background thread.

Here are some examples of this that have already been addressed in this PR:

* Previously, saving and loading a `TextBuffer` required its entire contents to be represented as a single JavaScript String. V8 has a hard limit on string length, so this would cause errors (#7210). Now, saving and loading are done in a background thread in C++, with minimal copying of data.
* When the file associated with an unmodified `TextBuffer` changes on disk, we diff the file's new contents with the buffer's contents in order to update the buffer in a way that preserves marker positions. Previously, this diff computation caused unresponsiveness in large files (atom/text-buffer#153). Now the diff is performed in C++ and on a background thread.
* Whenever the application is idle, we serialize the state of the window and persist it, so that we can restore it even if Atom crashes or power is lost. Previously, persisting a large file's text would cause hangs, so we had to disable this safety feature for large files (#13583). Now, because of our new representation of the buffer, we are able to efficiently serialize *just* the outstanding changes, so we can re-enable the feature for arbitrarily large files.
* The [line-ending-selector](https://github.com/atom/line-ending-selector) needs to detect what type of line endings a `TextBuffer` uses. Because we support files with *mixed* line endings, this entails scanning the entire buffer. Previously, this would take upwards of 50ms for a large file. Now, the search is performed asynchronously so that it causes no noticeable lag.

In addition to these improvements, this PR will make it very easy to optimize other features. Here are some follow-up PRs that I have in mind:
* When changing a large amount of text at once, indexing the new text in the `DisplayLayer` takes a long time. This is one of the main reasons that pasting a huge amount of text is slow. Some of this character-by-character computation could probably be done much faster in C++.
* Currently, using the [find-and-replace](https://github.com/atom/find-and-replace) package in a very large file can cause hangs when searches produce more than a hundred thousand results. We will now be able to run searches entirely in a background thread, eliminating any lag.
* Populating the [autocomplete-plus](https://github.com/atom/autocomplete-plus) package's `SymbolStore` can take a very long time for large files, so we had to [disable it](https://github.com/atom/autocomplete-plus/pull/782) for large files. Fuzzy matching should be able to be performed in C++ on a background thread, eliminating this restriction.

### Implementation details
<details>
<summary>Our new representation of the buffer:</summary>
<br/>

Suppose I create a `TextBuffer` with some initial text and then make three edits:

```js
const buffer = new TextBuffer('abc\ndefg\nhijkl')
buffer.setTextInRange([[0, 2], [0, 3]], "C")
buffer.setTextInRange([[2, 0], [2, 1]], "H")
buffer.setTextInRange([[1, 1], [1, 2]], "E")
assert.equal(buffer.getText(), 'abC\ndEfg\nHijkl')
```

This buffer is now represented internally like this:

![graph-1](https://cloud.githubusercontent.com/assets/326587/26644547/f3bbc95a-45e9-11e7-91a4-ceee77e90ecc.png)

The buffer is composed of two *layers*. The first layer stores the buffer's initial text as a flat, immutable string. This makes freshly-opened files very compact in memory. The second layer stores the unsaved edits in an object that we call a `Patch`. The `Patch` is structured as a [splay tree](https://en.wikipedia.org/wiki/Splay_tree) so that mutations and queries are very efficient, especially in areas that have recently been accessed.

Then, suppose I begin some asynchronous operation like saving the buffer to disk, or searching 
the buffer for a regex. Then, *while* that operation is still happening, I make some *additional* edits:

```js
const savePromise = buffer.save()
buffer.setTextInRange([[0, 0], [0, 1]], 'A')
buffer.setTextInRange([[2, 4], [2, 5]], 'L')
assert.equal(buffer.getText(), 'AbC\ndEfg\nHijkL')
```

At that point, the buffer's internal state would look like this:

![graph2](https://cloud.githubusercontent.com/assets/326587/26644968/63dd6562-45eb-11e7-9d66-9dfe94d243d9.png)

There are now *three* layers. Because the asynchronous `save` is still in progress, further changes that we make to the buffer are saved in a *new* Patch. When the `save` completes, we flatten the text down into one layer again.

</details>

### Effect on packages

I don't believe making `save` async has affected any of our bundled packages' functionality, though it has caused some test failures that I've had to fix:
* [x] https://github.com/atom/autocomplete-plus/pull/852
* [x] https://github.com/atom/whitespace/pull/155
* [x] https://github.com/atom/line-ending-selector/pull/45
* [x] https://github.com/atom/symbols-view/pull/222
* [x] https://github.com/atom/status-bar/pull/193
* [x] https://github.com/atom/tabs/pull/448

Depends on https://github.com/atom/text-buffer/pull/225
Depends on https://github.com/atom/superstring/pull/5
Depends on https://github.com/atom/node-pathwatcher/pull/121
Depends on https://github.com/atom/whitespace/pull/157

Closes https://github.com/atom/atom/issues/7210
Closes https://github.com/atom/atom/issues/11554
Closes https://github.com/atom/atom/issues/11012

maxbrunsfeld committed:
Start work on integrating new native TextBuffer version

maxbrunsfeld committed:
:arrow_up: autocomplete-plus

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
Avoiding writing to fixture files in Workspace.replace tests

maxbrunsfeld committed:
:arrow_up: text-buffer, whitespace, line-ending-selector

maxbrunsfeld committed:
Immediately fetch new buffers' git status in GitRepository

We used to emit buffer added events before the buffers were loaded, so we
would end up fetching their git status soon afterwords, when they emitted
the did-reload event.

maxbrunsfeld committed:
Wait for save to complete in pending state spec

maxbrunsfeld committed:
:arrow_up: symbols-view

maxbrunsfeld committed:
Don't create duplicate buffers if a path is opened twice concurrently

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
:arrow_up: status-bar

maxbrunsfeld committed:
Merge branch 'master' into mb-use-native-text-buffer

Commit:
Merge branch 'master' into mb-use-native-text-buffer

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld commented:
@atom/maintainers I think this PR is ready for others to test.

@ungb @Ben3eeE Some areas that could use testing:

Loading and saving buffers. This includes actual I/O as well as encoding/decoding. Areas in which I could imagine regressions:
  * Opening a file on a slow drive (e.g. `sshfs` and other network file systems)
  * Opening files in non-UTF8 encodings
  * Changing files on disk while they are open in Atom
  * Handling errors when reading/writing files (e.g. permissions issues)
  * Lots of simultaneous IO operations (e.g. multiple saves at the same time, multiple on-disk changes in rapid succession)

Working with buffers with large numbers of unsaved edits. All the operations that I've tested perform as well or better than before, but because of the new buffer representation, I could imagine some operations getting slower. We need to make sure that doesn't happen. Some types of operations I'm thinking of:
  * Performing a regex replace with lots of search results
  * Indenting large chunks of the buffer
  * Pasting large amounts of text
  * Normal editing *after* doing any of the above

32-bit platforms. I did some testing on a 32-bit linux VM, but possibly not enough.

maxbrunsfeld commented:
Download links for convenience:

* [atom-mac.zip](https://4010-3228505-gh.circle-artifacts.com/0/Users/distiller/atom/out/atom-mac.zip) from Circle
* [atom-windows.zip (x64)](https://ci.appveyor.com/api/buildjobs/82qxfouodneo0joq/artifacts/out%2Fatom-windows.zip) from Appveyor

maxbrunsfeld committed:
Fix handling of .save and .saveAs rejections

* Make Pane.close, Pane.saveActiveItem, and Pane.saveActiveItemAs async.
* Refactor the logic for prompting to save on window unload

ungb commented:
Few things I noticed while testing this:

-  when you open a decently large file ~70mb and then save it, the heap increases by 10x and never gets freed.  Over time, this will eventually cause atom to crash. 
![image](https://cloud.githubusercontent.com/assets/5624870/26657743/a75f34da-461a-11e7-938b-196ed074a8a1.png)
This doesn't repro in beta, talking to @maxbrunsfeld seems to do with caching that was occurring? 
Also, this doesn't seem to occur on smaller files.  Will investigate when this occurs

- Saving also seems to take a bit longer

- File sometimes blanks out.

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
:arrow_up: tabs

maxbrunsfeld committed:
Merge branch 'master' into mb-use-native-text-buffer

maxbrunsfeld committed:
Convert pane spec to JS

maxbrunsfeld committed:
:arrow_up: pathwatcher (prerelease)

maxbrunsfeld committed:
Merge branch 'master' into mb-use-native-text-buffer

maxbrunsfeld committed:
Use async/await in Pane spec

maxbrunsfeld committed:
Fix handling of aborted save in Pane.saveItem

maxbrunsfeld committed:
Fix Pane.close test

maxbrunsfeld committed:
:arrow_up: text-buffer and pathwatcher (prerelease)

maxbrunsfeld committed:
Remove code for preserving cursor position on buffer reload

Updating the markers correctly is handled by TextBuffer. This
was actually breaking it.

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
Fix unintended unload prevention for spec windows

maxbrunsfeld commented:
Ok, we've found and addressed a number of bugs in the last two days. Download links for the latest version:

* [mac](https://4070-3228505-gh.circle-artifacts.com/0/Users/distiller/atom/out/atom-mac.zip)
* [windows x64](https://ci.appveyor.com/api/buildjobs/5ifu2mgmwgjp5gdg/artifacts/out%2Fatom-windows.zip)

Glavin001 commented:
I notice you mentioned "making save async". Will there be any new Atom APIs allowing package developers to tell Atom to wait for packages to perform async operations to TextBuffer before saving? See https://discuss.atom.io/t/best-way-to-change-the-texteditor-textbuffer-on-save-when-my-function-is-asynchronous/43531 for more context to my question. Thank you!

lee-dohm commented:
@Glavin001 The `TextBuffer.onWillSave` event will continue to work the way it always has. I've tested it out using my [tabs-to-spaces package](https://atom.io/packages/tabs-to-spaces) on a [very large file](https://github.com/lee-dohm/tabs-to-spaces/blob/master/sample/jquery-git2.js.txt).

Glavin001 commented:
>  The TextBuffer.onWillSave event will continue to work the way it always has.

Therefore this PR may not solve my issue [as suggested](https://discuss.atom.io/t/best-way-to-change-the-texteditor-textbuffer-on-save-when-my-function-is-asynchronous/43531/3?u=glavin001). Hmm..

I'll elaborate, and hopefully you have some ideas, @lee-dohm. My problem has been `TextBuffer.onWillSave` does not wait for my asynchronous operations to complete before actually saving the buffer.

My situation with [Atom-Beautify](https://github.com/Glavin001/atom-beautify) is similar to the following:

```
const buffer = editor.getBuffer();
buffer.onWillSave(() => {
    // Please wait for async function to finish
    setTimeout(() => {
       // Nope! Buffer has already been saved!
       buffer.setText('Async change completed!');
    }, 1000);
})
```

Looking at the [`TextBuffer.onWillSave` implementation](https://github.com/atom/text-buffer/blob/v11.4.1/src/text-buffer.coffee#L347-L348) I think this only works when the operation is synchronous. *I would like to be able to return a promise (or callback) to let Atom know I have completed my changes and Atom can now save.* If this feature is also desirable to the Atom team, I'd be happy to submit a Pull Request, if I could have a little guidance 😄 .

maxbrunsfeld commented:
> My problem has been TextBuffer.onWillSave does not wait for my asynchronous operations to complete before actually saving the buffer.

@Glavin001 Thanks for the explanation. This problem comes up a lot. I don't think that I'll solve it in this particular PR, but it will become a lot *easier* to solve after this PR, because I've adjusted all of Atom's codebase to expect `save` be asynchronous instead of synchronous. So now it will be a pretty small change to allow `onWillSave` callbacks to delay the save operation.

joefitzgerald commented:
@Glavin001 I also have a need for the `onWillSave` event to support asynchronous functions (likely returning a promise).

I want to run `gofmt` or `goimports` for people writing Go code. Today this is synchronous, which leads to a sub par user experience as an external process in invoked.

In general, code formatters need to run at this time; but in a way that protects the user in the event of some failure.

maxbrunsfeld commented:
@ungb I've tracked down the problem where you found where a large buffer was 'blanked out' after its file was written to externally. I'll have a fix shortly.

Glavin001 commented:
Thank you for your feedback, @maxbrunsfeld ! This is great to hear. I would love to contribute and bring this next change into reality, if someone else is not already planning to. How can I help? 😃

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

ungb commented:
found https://github.com/atom/atom/issues/14722 while testing this branch.  Will pull from latest and try again.  Was able to repro on windows and mac.

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
:arrow_up: whitespace

maxbrunsfeld committed:
Merge remote-tracking branch 'origin/master' into mb-use-native-text-buffer

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
Don't serialize undo/redo stacks when the user is idle

maxbrunsfeld committed:
:arrow_up: whitespace (prerelease)

maxbrunsfeld committed:
:arrow_up: line-ending-selector

maxbrunsfeld committed:
Fix exception on cancelling path selection when saving

Fixes #14722

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
Use a more recent c++ toolchain on travis

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

maxbrunsfeld committed:
:arrow_up: text-buffer (prerelease)

t9md commented on src/pane.coffee:
```diff
@@ -603,55 +605,71 @@ class Pane
-    @destroyItem(item) for item in @getItems()
-    return
+    Promise.all(
+      @getItems().map(@destroyItem.bind(this))
```
@maxbrunsfeld 
Noticed while I read code, using `bind` here is bad.
Since `Array.prototype.map` pass item index as 2nd arg, which goes to `@destroyItems` which treat 2nd arg as `force` switch. Which is not intentional I believe.

maxbrunsfeld replied:
Wow, thanks for catching that @t9md! Are you interested in opening a PR fixing that bug?

t9md replied:
Yes, I will try tommorrow.

t9md commented on src/pane.coffee:
```diff
@@ -603,55 +605,71 @@ class Pane
+    Promise.all(
+      @getItems()
+        .filter((item) => item isnt @activeItem)
+        .map(@destroyItem.bind(this))
```
here too.

maxbrunsfeld commented:
@Glavin001 @joefitzgerald The `TextBuffer.onWillSave` hook now allows you do perform async work before the buffer is saved. You just need to return a promise from the event handler. For example:

```js
buffer.onWillSave(async () => {
  await somethingSlow()
})
```

See https://github.com/atom/text-buffer/pull/254.

This change has been released in Atom v1.21.0-beta1.

Glavin001 commented:
Awesome! Thank you, @maxbrunsfeld!
